### ISSUE TITLE
Feedback on 1.5.1b1: Saving/Updating passwords

### ISSUE THREAD

**Issue by dlech** [luckyrat/KeeFox#483]
I just logged into my local router (dd-wrt). KeeFox filled in the password correctly (http auth). However, after I clicked OK, KeeFox thought this was a new password and asked me if I wanted to save. From what I understand of the release notes, this is to be expected sometimes. So I clicked yes, to save the password again. So trying to do this, there are some issues.

Here is what it looks like.
![selection_025](https://cloud.githubusercontent.com/assets/963645/8136863/e650793a-1104-11e5-90ca-b40c01aff4d0.png)

First of all, it is not clear that "Create New Entry" and "Update Existing Entry" are tabs. I didn't realize this until I looked really closely and notice that the top corners were rounded and the bottom are not. This could use some more styling to make this obvious at first glance. I think using different colors for the tabs would be good so that they are not easily confused for buttons. Also, if you can make the vertical sides of the tabs at a slight angle, that would make them more tab-like. And finally, a stronger border and different background color for the tab content area would help it stand out more.

The X button is kind of ugly too, but that is minor. I would suggest a red background with a white X, more rounded corners and center the X vertically and horizontally.

So my first impression was that I was hesitant to click "Create New Entry" or "Update Existing Entry" because I was it looks like a button and I was afraid that it might do something that I did not want to happen. After I got over my fear, I clicked on the "Update Existing Entry" tab and searched for the existing entry that I wanted to update. The result is this:

![selection_027](https://cloud.githubusercontent.com/assets/963645/8137438/0d373788-1109-11e5-856e-71ff34475d24.png)

I don't find this intuitive either. The first time I saw it, I just started typing in the search box because I did not realize that it had already searched for the Current Website. It happens that the hostname is the same as my email address, so it pulled up lots of entries - any one that uses my email as a user name.

Now, if I type something in the search box, does it search for my term AND the hostname of the current website? If I select "All Websites" does it search only entries in my database that have a URL or does it search all entries in the database. Ideally, you should be able to search all entries in the database so that you can update any existing entry. It is very rare that I add a URL when creating new entries manually.

OK, so by searching, I have found the entry that I want to update and click on the entry. I am surprised that I get a whole new "page", if you will:

![selection_024](https://cloud.githubusercontent.com/assets/963645/8136801/5c19539a-1104-11e5-9790-868909235276.png)

What are all the "Change" buttons for? If I want to update my entry, do I click "Change" or do I click "Save"? Or can I change each of the individual lines?

I was brave and nervously clicked on each of the "Change" buttons. The first takes me back to search. OK. so far so good. The second throws a massive amount of text at me. The third give a bunch of radio button options. Yikes!

![selection_028](https://cloud.githubusercontent.com/assets/963645/8137780/60a85c9c-110b-11e5-96cc-bc93327513fc.png)

I know more about the internal workings of KeeFox that the average person, so after pondering this a bit, I think I understand what it is saying, but it's a lot to think about and all I wanted in the first place is for KeeFox to not nag me about a password that it already knew enough about to fill in the HTTP auth box.

KeePass has history enabled by default so just replace everything and don't ask me. If there is an existing URL, then add the new one to the "extra" URLs on the KeeFox tab in KeePass rather than replacing the existing URL. This is the third option above.

So what if you rearrange this window a bit. Maybe like this:

![selection_026](https://cloud.githubusercontent.com/assets/963645/8137293/e8e733de-1107-11e5-9d90-ab67ba41398d.png)

![selection_029](https://cloud.githubusercontent.com/assets/963645/8138099/c1775314-110d-11e5-8b87-fe97e3a0cd7e.png)

Key points here:
* More borders/frames to separate things
* Buttons have more descriptive words to indicate what they do
* More user-centric language
* Don't ask so many questions/give too many options
* The search box is pre-filled with the hostname of the current website. The text should also be selected if possible so that the user can just start typing to replace it.
* Search always searches the entire database(s), not just entries with URLs. May need a way to indicate which database entries are in for those with more than one database.

**Comment by luckyrat** [comment]
Thanks for all the feedback!

I think that before I get into the detail of your suggestions, I should just post this quick update to give you food for thought...

The images below are what I have designed and what has appeared on each version of Windows I have tested. Suffice to say that there must be something very odd going on with the default CSS on whichever OS you're using in order to end up with such a different result!

![savepassword1](https://cloud.githubusercontent.com/assets/1211375/8139649/284813a0-114d-11e5-95dc-26b7570ac541.png)
![savepassword2](https://cloud.githubusercontent.com/assets/1211375/8139650/284bcbb2-114d-11e5-8fc2-9c2785664e82.png)
![savepassword3](https://cloud.githubusercontent.com/assets/1211375/8139652/284f46d4-114d-11e5-85a3-22387c8277a0.png)
![savepassword4](https://cloud.githubusercontent.com/assets/1211375/8139651/284dca02-114d-11e5-8c8c-751eb08d52db.png)

**Comment by dlech** [comment]
I am using Elementary OS Freya, which is essentially Ubuntu 14.04 with a different desktop environment. I can test on regular Ubuntu to see if it is the same.

Also I just got this:

![selection_030](https://cloud.githubusercontent.com/assets/963645/8140606/b5e122b8-1122-11e5-8ae6-95d3d2f2722b.png)

**Comment by luckyrat** [comment]
> it is not clear that "Create New Entry" and "Update Existing Entry" are tabs

Hopefully this is clearer in the intended/windows design?

> I would suggest a red background with a white X

I tried this initially but it felt a bit "OMG! What have I done wrong why is there is a big red thing grabbing my attention?". When we work out how to get the styles to work on other OSes you can see the effect because I retained it as the hover style. Any further thoughts?

> I don't find this intuitive either. The first time I saw it, I just started typing in the search box because I did not realize that it had already searched for the Current Website.

Well, this is tough because you only get one first impression but can you estimate whether you would find the "correct" style more intuitive? (Incidentally, it should look identical to the search interface on the main panel but maybe that is knackered on non-Windows machines too?)

> does it search for my term AND the hostname of the current website?

Yes that's the idea (assuming you keep with the "current website" option in the select dropdown) but the hostname search is limited to the URL fields so it shouldn't be picking up anything that matches purely on username (unless you type the hostname in manually which I think is what you said you did).

> Ideally, you should be able to search all entries in the database so that you can update any existing entry.

Yeah that isn't how it works at the moment but I've been contemplating a change so that KeePassRPC returns all entries and we just filter out the ones without URLs in KeeFox at the points where URLs are required (primarily the list of logins and main panel search results). It's been a low priority until recently but with the new update functionality we definitely have another good reason to investigate the impact of that change.

I imagined the update password feature would be most used by people updating passwords that they had already saved (or at least used) through KeeFox so the lack of entries without URLs shouldn't be a problem for that (majority?) use-case. But easing the assignment of URLs to pre-existing KeePass entries is certainly a valid thing to try to do with the feature so it's definitely worth thinking if we can enable that use-case too.

> I am surprised that I get a whole new "page"

I originally intended for this to transition neatly (maybe slide to the left) but haven't prioritised that work for the initial beta on the assumption that it would be OK to release without it. Perhaps the jarring / surprise you felt would have been lessened with a smooth transition? Or perhaps it's just that you thought you were done and weren't!

> What are all the "Change" buttons for?

OK, so there might be a couple of extra pixels of vertical padding on my design but overall I think you got an accurate first impression of this page but let me know if you've changed your mind after seeing the correct design.

I originally imagined these as little links directly next to the text of each line but it just looked too cluttered so I pulled them all off to the right to make it pretty. It looks like that has caused too much of a disconnect between the button and the thing it is going to affect. Maybe extra vertical padding will be enough? Or some kind of background or border around each of the 3 rows?

I'm not sure if there is another word we could use that would correctly indicate that it's a local change operation rather than for the entire "change password" operation but I am open to suggestions.

The reasoning behind the extra URL options here is kind-of already borne out by your comment that you would usually want option 3 for URL modifications - I'd normally want option 1 :-) ... although not always. I thought that would be the case for most people but maybe there could be benefit for an option to change the default selection.

As for the middle "change" button, I think it is important that people know what it is that will be updated and are given an option to do it manually if it's not the type of update they want.

The intention behind all the "change" buttons is that there should rarely be any need to click them and that if in doubt, ignoring them and just clicking the primary "Save" button gets you what you expected.

> just replace everything and don't ask me

We can't accurately replace things like regex and blocked URLs. On reflection, the bits we don't replace may actually be more of an edge-case than when I first designed the interface so perhaps we could alter the middle "change" option... but I do like the idea of letting people get to the entry in KeePass easily and if we have to have the URL options anyway, I'm not sure how much of a benefit getting rid of that middle change option will be. Worth considering in more depth though I think.

> Buttons have more descriptive words to indicate what they do

Good call. I was lazy by just using a single button to do everything but updating the "Save" button text to say "Create" or "Update" would be a definite improvement.

> The search box is pre-filled with the hostname of the current website

That should be unnecessary provided that the "Current website" default search filter is working correctly (since the styles were confusing when you last tried it, I'm not sure if this still needs addressing).

> May need a way to indicate which database entries are in for those with more than one database.

I think that should already be represented in the tooltip which I thought would be good enough for those rare occasions where there are naming conflicts across databases?

> Also I just got this:

Any idea what might have triggered that?

I've seen a similar problem when trying to save more than one password concurrently (in different tabs) but haven't had time to look into it in detail (was hoping I'm the only one that would try something so crazy!)

I'll check back on this tomorrow so let me know if you have any further comments or suggestions. Thanks!

**Comment by dlech** [comment]
> Hopefully this is clearer in the intended/windows design?

Yes it is. Although I still think it could benefit from some wording that indicates that these are the "yes" options, like in my mock up where it says "Yes, I want to...".

> I tried this initially but it felt a bit "OMG! What have I done wrong why is there is a big red thing grabbing my attention?". When we work out how to get the styles to work on other OSes you can see the effect because I retained it as the hover style. Any further thoughts?

How about borrowing the X from the tabs. Normally, they look like this:

![selection_031](https://cloud.githubusercontent.com/assets/963645/8142037/34cd5df4-1136-11e5-805c-f9613831f9f6.png)

And when you hover, they turn red...

![selection_035](https://cloud.githubusercontent.com/assets/963645/8142050/a04e6910-1136-11e5-9183-965749c25693.png)

Small and familiar.

> Well, this is tough because you only get one first impression but can you estimate whether you would find the "correct" style more intuitive? (Incidentally, it should look identical to the search interface on the main panel but maybe that is knackered on non-Windows machines too?)

No, the style really didn't affect my impression in this regard. It would be more clear (and perhaps a bit too verbose) to say "Entries with URL matching www.example.com" instead of "Current website" and "All entries" instead of "All websites". You could possibly add a 3rd option as well "All entries with a URL".

The fact that KeeFox only uses entries with URLs and ignores those without is something that took me quite a while to grasp when I first started using KeeFox. So the more plainly you spell this out, the better.

> I originally intended for this to transition neatly (maybe slide to the left) but haven't prioritised that work for the initial beta on the assumption that it would be OK to release without it. Perhaps the jarring / surprise you felt would have been lessened with a smooth transition? Or perhaps it's just that you thought you were done and weren't!

It was the "I thought I was going to just click the entry and then click save next and that would be it but instead I now have to read a whole bunch more stuff and there are more buttons" that surprised me.

> I originally imagined these as little links directly next to the text of each line but it just looked too cluttered so I pulled them all off to the right to make it pretty. It looks like that has caused too much of a disconnect between the button and the thing it is going to affect. Maybe extra vertical padding will be enough? Or some kind of background or border around each of the 3 rows?

A horizontal line between each section might be enough. Putting each section in its own frame is probably better. (I'm using "frame" in the Windows Forms sense, not sure if it is called the same thing in XUL).

> I'm not sure if there is another word we could use that would correctly indicate that it's a local change operation rather than for the entire "change password" operation but I am open to suggestions.

For the first line, say "Search again" or "Back to search". For the second line, say "More info". For the third line, say "More options"

> The reasoning behind the extra URL options here is kind-of already borne out by your comment that you would usually want option 3 for URL modifications - I'd normally want option 1 :-) ... although not always. I thought that would be the case for most people but maybe there could be benefit for an option to change the default selection.

If you must offer all of these options, then make the button say "More options" as suggested above. Also, save the selection in preferences and use that as the default option. Don't make me go into the KeeFox options dialog to set this, just remember the last one I selected.

> As for the middle "change" button, I think it is important that people know what it is that will be updated and are given an option to do it manually if it's not the type of update they want.

For me, the expected behavoir here is for KeeFox to write over my existing entry just as if it was a new entry. By explaining it here, it just makes me overthink it.

At the very least though, I would make the button say "Don't update, open the entry in KeePass instead". And I think I would make the button visible by default, not hidden until you click another button.

> That should be unnecessary provided that the "Current website" default search filter is working correctly (since the styles were confusing when you last tried it, I'm not sure if this still needs addressing).

As suggested above, this would be more clear if this was spelled out more plainly. "Current website" just does not at all make me think "entries that have a URL that matches the hostname of the current website"

> I think that should already be represented in the tooltip which I thought would be good enough for those rare occasions where there are naming conflicts across databases?

I don't use multiple databases, so you should probably rely on feedback from others here. I don't see a database name in the tooltip though - perhaps because I only have one database open.

> Any idea what might have triggered that?

It seems like the "Do you want KeeFox to save this password?" boxes "pile up" if you don't close them using the X. I found that I could close any one of the 3 and I would have 2 left, but then I could not close the other 2. If I went to a site that caused the 3rd box to appear again, then I could again close any one of the 3, but the other 2 would not close.

**Comment by dlech** [comment]
Regarding the styling (or lack thereof), I am seeing the expected styling when I manually click on the KeeFox icon/button. It seems that when the popover is triggered from some other source, that is when I don't have the proper styling.

I tried this on vanilla Ubuntu 15.04 as well and go the same results - that is, with all other addons, including Ubuntu integrations addons, disabled.

**Comment by luckyrat** [comment]
> Yes it is. Although I still think it could benefit from some wording that indicates that these are the "yes" options, like in my mock up where it says "Yes, I want to...".

OK well let's see if we get further feedback about this during beta testing since I don't want to add extra words if it's unnecessary.



> How about borrowing the X from the tabs.

I thought they might be too small. I dunno, I think the current design strikes a good balance between making sure people know it's there and what it does but without being too "in your face".


> It would be more clear (and perhaps a bit too verbose) to say "Entries with URL matching www.example.com" instead of "Current website" and "All entries" instead of "All websites". You could possibly add a 3rd option as well "All entries with a URL".

I'm trying to keep consistency with the search feature in the main panel wherever possible. In the context of only ever returning results that have a URL, I think "All websites" makes more sense to a user, especially a relatively new one or non-technical person - they definitely know what a "website" is but I'm not sure whether "entry" is going to be quite as widely understood.

However, if we implement #484 it might be worth swapping out the "All websites" option for a 3rd option, simply named "All entries". Since we default to "Current website", I think the common case of updating an existing entry is covered without introducing the user to any KeePass specific terminology ("entry") and anyone brave enough to change the select drop down option would probably at least get a vague enough understanding that they are "widening search to include more things" which is probably good enough?

I took a similar decision regarding "Current website" - all users will understand what that text refers to, even if it is a little abstracted from the actual search query behind the scenes. As you say, "Entries with URL matching www.example.com" is a more accurate description of what entries are available for searching but there are a few problems with that approach:

1. As you say, it is too verbose - even more so when you consider:
  1. other languages can add up to 50% of length compared to English
  2. the situation with logins contained in frames is not entirely resolved yet and it is likely that we'll end up searching all subframes too so the actual domain being searched could end up being a long list. See #485.
2. "URL" is not as widely understood as "website address" and that in turn is less understood than "website".
3. As I mentioned earlier, "entry"/"entries" is a concept that might not be understood until the user has been using KeeFox or KeePass for a little while.

> The fact that KeeFox only uses entries with URLs and ignores those without is something that took me quite a while to grasp when I first started using KeeFox. So the more plainly you spell this out, the better.

Fair point, especially for the majority of active users that come from a "KeePass first" background. I just want to keep this interface as clear as possible so if we can shift the explanation of "entries without URLs are ignored" to somewhere else (or implement #484) that would be my preferred option. In fact, I'm working on the getting started tutorial at the moment which is one further place we could explain this (but I won't until we've decided about #484). I'll let you know as soon as I have something public for you to look at but the main change relevant to this discussion is that I'm planning to replace the section about configuring an alternative URL with a bit about updating a password so we could put an occasional explanatory point in that section (although obviously I still want the interface to be reasonably easy to understand with no prior experience of KeeFox, KeePass or even password managers in general).

> A horizontal line between each section might be enough. Putting each section in its own frame is probably better. (I'm using "frame" in the Windows Forms sense, not sure if it is called the same thing in XUL).

OK I'll have a play around with that.

> For the first line, say "Search again" or "Back to search". For the second line, say "More info". For the third line, say "More options"

Great suggestions.

> Also, save the selection in preferences and use that as the default option. Don't make me go into the KeeFox options dialog to set this, just remember the last one I selected.

What if I only wanted to use a different option occasionally? Do you think that offering a checkbox along the lines of "Remember this choice" would be sensible?

Which then gets me thinking...

> For me, the expected behavoir here is for KeeFox to write over my existing entry just as if it was a new entry. By explaining it here, it just makes me overthink it.

That's my expectation too, but I'm not certain if that matches everyone's expectation. Maybe some people would expect that only the value of the fields they edited would be changed? I.e. if their entry contained some fields that were not submitted to the current form, would they then be surprised when those fields have been deleted by KeeFox?

Maybe I am overthinking it and in fact we should lean more heavily on the KeePass history feature to allow recovery from unexpected data loss.

... which would therefore lead to there being a reduced need for the "Don't update, open the entry in KeePass instead" option. I'm still on the fence about the benefit of that, partly because the benefit might increase if #484 results in this interface being the only part of KeeFox that can reveal entries without URLs (but then how likely is it people will want to do manual editing of those entries? Probably very unlikely.)

... which would lead to a potential solution that takes into account these various improvements:

![updatepasswordsketch](https://cloud.githubusercontent.com/assets/1211375/8144261/033e238a-11cd-11e5-9a3c-4c980a8a03ad.png)

That gets displayed at least once to every user but any that are interested/bothered enough to explore "more options" can be rewarded with the previous 4 radio buttons but also a checkbox to "Remember this choice". Once that choice has been remembered, it can be altered through the KeeFox options panel - per #486.

I might still use the text explaining what gets updated by KeeFox but perhaps relegate it to the tutorial or a manual page.

> As suggested above, this would be more clear if this was spelled out more plainly. "Current website" just does not at all make me think "entries that have a URL that matches the hostname of the current website"

I guess we need to get to the bottom of the seriousness of this disconnect between the abstraction of "current website" and the actual entries that are available for search. For a start, would it even matter at all if we assume that #484 has been implemented? If we don't do #484, then is the impact limited to an initial frustration for the first time only? And is that frustration more related to the fact that the feature does not allow you to find the entries you are looking for rather than the imprecision of the feature labelling?

> I don't use multiple databases, so you should probably rely on feedback from others here. I don't see a database name in the tooltip though - perhaps because I only have one database open.

I'll double check this.

> It seems like the "Do you want KeeFox to save this password?" boxes "pile up"

Yeah that's the exact behaviour I saw when saving concurrently in multiple tabs. The old notifications should definitely be removed each time a new one is displayed (unless it is for a different tab) so it looks like the multi-tab saving problem is more generic which does increase the importance of fixing it but hopefully also means it'll be easier to find the solution! Will track in #487

> Regarding the styling (or lack thereof), I am seeing the expected styling when I manually click on the KeeFox icon/button. It seems that when the popover is triggered from some other source, that is when I don't have the proper styling.

The popover triggered following a form submission is a XUL panel element whereas the one that comes up when you click the KeeFox button is an Australis-related XUL panelview element. It looks likely that the differences between those two types of popover panel are more significant on Linux than Windows.

> I tried this on vanilla Ubuntu 15.04 as well and go the same results - that is, with all other addons, including Ubuntu integrations addons, disabled.

Well that's good news from a reproducibility perspective :-) I won't focus on this for the time being though since I'd rather get any string changes and behaviour changes done first and worry about why the #keefox-persistent-panel element styling is messed up while the translators are working on the new strings.

**Comment by dlech** [comment]
> I think the common case of updating an existing entry is covered without introducing the user to any KeePass specific terminology ("entry") and anyone brave enough to change the select drop down option would probably at least get a vague enough understanding that they are "widening search to include more things" which is probably good enough?

Yes, I suppose that is good enough. On the other hand, I think the term "password entry" should be rather easily recognized.

> What if I only wanted to use a different option occasionally? Do you think that offering a checkbox along the lines of "Remember this choice" would be sensible?

If it is only occasionally, then it is not really a big deal to switch it back to the preferred choice on the next go around. Adding a checkbox on the other hand just makes one more thing that you have to think about and slow you down.

> I.e. if their entry contained some fields that were not submitted to the current form, would they then be surprised when those fields have been deleted by KeeFox?

Yes, this would be surprising in a bad way. Furthermore, the current explanation did not get this across to me. I think you should specifically mention that all form fields, including other pages of multi-part forms, will be deleted.

> Maybe I am overthinking it and in fact we should lean more heavily on the KeePass history feature to allow recovery from unexpected data loss.

This seems to be the direction that "modern" apps are going. [[1]] Don't ask "are you sure?", just do it - but make sure that the user can undo it.

[1]: https://elementary.io/docs/human-interface-guidelines#always-provide-an-undo

> which would lead to a potential solution that takes into account these various improvements:

I like it. Although I would still separate the Create/Update button and put one on each tab so that is is grouped in with the other user input for its specific task.

> Once that choice has been remembered, it can be altered through the KeeFox options panel

I would rather not have to go into the KeeFox options panel to change it. In fact, I don't even know if it would occur to me to look there.

> I guess we need to get to the bottom of the seriousness of this disconnect between the abstraction of "current website" and the actual entries that are available for search. For a start, would it even matter at all if we assume that #484 has been implemented? If we don't do #484, then is the impact limited to an initial frustration for the first time only? And is that frustration more related to the fact that the feature does not allow you to find the entries you are looking for rather than the imprecision of the feature labelling?

Yes, if there was an option for "All password entries" (#484), then that would be a very strong hint that "All websites" is a limited set of entries and "Current website" is even more limited still. I would really like to see #484 implemented, so I'm just going to assume that this is the case and not spend my time pondering the other possibilities.

> The popover triggered following a form submission is a XUL panel element whereas the one that comes up when you click the KeeFox button is an Australis-related XUL panelview element. It looks likely that the differences between those two types of popover panel are more significant on Linux than Windows.

Is there something preventing the panelview from being used in both cases?

**Comment by luckyrat** [comment]
> Adding a checkbox on the other hand just makes one more thing that you have to think about and slow you down.

I agree if it weren't for the possibility of entirely removing that 2nd stage of the update process.

> I would rather not have to go into the KeeFox options panel to change it. In fact, I don't even know if it would occur to me to look there.

I think putting it in the options panel is the only way that we'd be able to remove the 2nd stage from the workflow and therefore achieve the "one click on an entry in the search results + one click on the Update button" experience you were expecting initially.

> I think you should specifically mention that all form fields, including other pages of multi-part forms, will be deleted.

I did consider that but it resulted in a chunk of text that was either making unrealistic assumptions about the user's knowledge of advanced features (most won't ever use a multi-page form before updating a single-page password) or so large that it is likely to be skimmed over and not fully understood anyway. I thought the term "replaced" managed to convey the implications succinctly but it appears not.

Incidentally, #411 would give us an indication of the page number we're saving but it will be impossible to guess correctly every time and in the context of updating a password I think we'd still need to acknowledge the risk that multi-page forms will be broken by saving a single page so I don't think it'll be much help here.

> This seems to be the direction that "modern" apps are going. [1] Don't ask "are you sure?", just do it - but make sure that the user can undo it.

Yeah without question that is the approach I would always prefer to take but replicating the undo functionality within KeeFox itself would be complicated and probably not very useful (the user is unlikely to realise something has gone wrong until the next time they use the website possibly days or weeks later). The KeePass history functionality does offer us a type of undo functionality but in my opinion it is too disconnected from the initial "save" operation to be able to assume all users will find it in a timely fashion. Hence why my initial approach erred on the side of excessive warnings ala "are you sure?".

> Although I would still separate the Create/Update button and put one on each tab so that is is grouped in with the other user input for its specific task.

I agree that would be better; at the moment it's the way it is because I was concerned about the extra height it would take up in a location that can't be scrolled and it also allows me to reuse code from the rest of the notifications system, but if I have enough time I'll probably make this change too. It might even be easier than changing the text on the existing button but I'll make that call when I'm implementing it.

> Is there something preventing the panelview from being used in both cases?

Yeah the panelview has lots of XUL magic around it so that it can work within the Firefox main menu as well as from the toolbar. It appears that this prevents us from forcing the panelview to remain open (which is required partly because it's too easy for the user to click elsewhere and prematurely dismiss the notification but mainly because every navigation event also causes the panelview to close so it is displayed for only a split second before the "login success" page is loaded).

Conversely, we can't use a standard panel within the Australis interface if we want users to be able to put KeeFox within the main Firefox menu (or when it automatically falls into the overflow menu when there isn't enough horizontal space).

**Comment by luckyrat** [comment]
[1.5.2b2](https://github.com/luckyrat/KeeFox/releases/tag/v1.5.2b2) has hopefully fixed the styles on Linux. I managed to more-or-less replicate what you were seeing on my Ubuntu 14.04 VM. The width of the dropdown selector is smaller than on windows but not as small as on your system. We get very limited influence over that control so it might not be possible to modify it further. The close button is better but still not quite right - I think it's not worth spending any more time on that now.

Changing the Save button label dynamically is going to be awkward so I'll probably separate and move them into the inner panel instead. However, we've run out of vertical space at the moment so until the state of the middle "warning/more info/abort" section is resolved, I'll leave it as it is.

I'm going to postpone the renaming of the "change" buttons until the 2nd beta because that will require translation work. I think that the fixed styling and extra separation between sections should be enough for beta testers to understand what's going on but if you strongly disagree we could always delay the 1st beta release (currently aiming for weekend of 20th June).

**Comment by luckyrat** [comment]
I have an idea to resolve the conflict of wanting to warn users about what happens and wanting to keep the interface simple (as per my previous mockup).

When a user updates any password entry, we'll display some extra information in a new notification containing a "don't show again" checkbox.

Here's a draft:

Password successfully updated.

KeePass has an "undo" feature which you can access from your KeePass entry. Click on the "History" tab and then use the "Restore" button to revert to an older copy of the entry.

The selected entry will have all username, password and other form fields replaced with the values you have just entered.

KeeFox will also update the website icon (favicon) and any HTTP Auth Realm information. More complex modifications will require you to manually edit the entry.

Multi-page login forms may also require manual editing.

[] Do not display this information again

And mockup:

![updatepasswordsketch2](https://cloud.githubusercontent.com/assets/1211375/8195683/7652cff4-147c-11e5-9dda-e7facad8f246.png)
What do you think?

**Comment by luckyrat** [comment]
The getting started tutorial has been updated - still needs a few screenshots and a final review once the work related to this issue has been done.

I've released the first beta (although not yet to the automatic update channel due to an AMO bug): https://github.com/luckyrat/KeeFox/releases/tag/v1.5.2b3

**Comment by dlech** [comment]
I've been busy with other things this last week. I'll try to set aside some time to look at this again this coming week.

**Comment by luckyrat** [comment]
OK I'm going for these tasks now:

- [ ] "Password successfully updated" notification (but with a button instead of checkbox since ticking a checkbox and then selecting the x to dismiss was a bit of a late-night odd idea!)
- [ ] Remove middle "more info" section from 2nd stage of update process
- [ ] Rename "search again" and "more options" buttons
- [ ] Split and migrate create/update button into the inner panel/frame

I'll leave the URL options behaviour as-is for the moment since I'm not certain of the right balance of defaults and persistent options yet.

**Comment by dlech** [comment]
I've just tried this again with the latest git (8905447cc02d235f0b20c192d59541f5846598fe) and the whole process is much smoother. It still gets across the information that you want without getting in my way or confusing me too much.

I think I can call this "good enough for me", so I'll go ahead and close this issue. I think we have already opened new issues for any other outstanding points in this discussion.
